proposals/0000-no-toplevel-field-selectors.rst

```diff
@@ -0,0 +1,250 @@
+
+Record construction/update syntax and pattern matching will work as before, the
+disambiguation handled by ``DuplicateRecordFields``. The necessary selectors
+will be part of a ``Record(..)`` export, or can also be named individually. They
+always have to be associated with the data type though, because there is no more
+toplevel selector (see `Example`_).
+
+A function for Template Haskell to go from a record field name to a selector for
+that field will be provided, because it's not possible anymore to go from field
+name directly to selector.
+
+A new ``TH`` function is added which takes a ``Name`` (for the constructor) and
+a ``String`` (for the field) and returns a ``Q Exp``.
+
+.. code-block:: haskell
+
+    mkFieldSelector :: Name -> String -> Q Exp
+
+Example
+^^^^^^^
+
+Given a data structure
+
+    data Foo = Foo { bar :: Int, baz :: String }
+
+The following will be available:
+
+- the type ``Foo``
+- the constructor ``Foo``
+- the two functions ``bar`` and ``baz``
+- the names ``bar`` and ``baz`` for record construction (``Foo { bar = 3, baz = "foo" }``)
+- the names ``bar`` and ``baz`` for ``RecordWildCards``
+
+If the language extension ``NoFieldSelectors`` is enabled for the module
+or ``Foo`` specifically, all of the above will be generated, except for the two
+functions ``bar`` and ``baz``.
+
+Wildcard exports will work as before, except for the two functions. Even if
+these functions are otherwise defined, the wildcard will not export them.
+Exporting the names for record construction now has to be specific to the
+record. Without ambiguitiy, previously this was equivalent
+
+.. code-block:: haskell
+
+    module A where (Foo(Foo, bar, baz))
+    data Foo = Foo { bar :: Int, baz :: Int }
+
+.. code-block:: haskell
+
+    module B where (Foo(Foo, bar), baz)
+    data Foo = Foo { bar :: Int, baz :: Int }
+
+Because of the new semantics, these two export statements are now different. The
+first one will export the field ``baz``, but not the function ``baz``, while the
+second one will export the function ``baz``, but not the field ``baz``. Because
+of this change, writing out all selector functions by hand is still different,
+because they all have to be exported manually.
+
+.. code-block:: haskell
+
+    {-# LANGUAGE NoFieldSelectors #-}
+    module A where (Foo(Foo, bar, baz))
+    data Foo = Foo { bar :: Int, baz :: Int }
+    baz = 42
+
+Which would be equivalent to:
+
+.. code-block:: haskell
+
+    {-# LANGUAGE NoFieldSelectors #-}
+    module A where (Foo(..))
+    data Foo = Foo { bar :: Int, baz :: Int }
+    baz = 42
+
+A second module, ``B``, which does not export the selector ``baz`` of
+constructor ``Foo``, but instead exports the toplevel bind ``baz``.
+
+.. code-block:: haskell
+
+    {-# LANGUAGE NoFieldSelectors #-}
+    module B where (Foo(Foo, bar), baz)
+    data Foo = Foo { bar :: Int, baz :: Int }
+    baz = 42
+
+The updaters can still be used when exported (as in module ``A``).
```

Comment:
I think you can just say the "field" can still be used. I've always understood "field" to refer to the occurrence of a field in record construction/update/pattern syntax, and "selector" to refer to the top-level projection function.

Reply:
I call them `RecordUpdater` in my GHC branch, hence the name.